common/types/src/contract.rs

```diff
@@ -0,0 +1,84 @@
+use alloc::string::String;
+use alloc::vec::Vec;
+use alloc::format;
+use codec::{Decode, Encode};
+
+pub use phala_mq::ContractId;
+
+pub type ContractId32 = u32;
+pub const SYSTEM: ContractId32 = 0;
+pub const DATA_PLAZA: ContractId32 = 1;
+pub const BALANCES: ContractId32 = 2;
+pub const ASSETS: ContractId32 = 3;
+pub const WEB3_ANALYTICS: ContractId32 = 4;
+pub const DIEM: ContractId32 = 5;
+pub const SUBSTRATE_KITTIES: ContractId32 = 6;
+pub const BTC_LOTTERY: ContractId32 = 7;
+
+/// Contract query request parameters, to be encrypted.
+#[derive(Encode, Decode, Debug)]
+pub struct ContractQuery<Data> {
+    pub head: ContractQueryHead,
+    /// The request data.
+    pub data: Data,
+}
+
+/// Contract query head
+#[derive(Encode, Decode, Debug)]
+pub struct ContractQueryHead {
+    /// The contract id.
+    pub id: ContractId,
+    /// A random byte array generated by the client.
+    pub nonce: [u8; 32],
+}
+
+/// Contract query response, to be encrypted.
+#[derive(Encode, Decode, Debug)]
+pub struct ContractQueryResponse<Data> {
+    /// The nonce from the client.
+    pub nonce: [u8; 32],
+    /// The query result.
+    pub result: Data,
+}
+
+pub struct Data(pub Vec<u8>);
+
+impl Encode for Data {
+    fn size_hint(&self) -> usize {
+        self.0.len()
+    }
+    fn encode_to<T: codec::Output + ?Sized>(&self, dest: &mut T) {
+        dest.write(&self.0)
+    }
+}
+
+/// Contract query error define
+#[derive(Encode, Decode, Debug)]
+pub enum ContractQueryError {
+    /// Signature is invalid.
+    InvalidSignature,
+    /// No such contract.
+    ContractNotFound,
+    /// Unable to decode the request data.
+    DecodeError,
+    /// Other errors reported during the contract query execution.
+    OtherError(String),
+}
+
+impl From<ContractQueryError> for prpc::server::Error {
+    fn from(err: ContractQueryError) -> Self {
+        Self::ContractQueryError(alloc::format!("{:?}", err))
+    }
+}
+
+pub fn id256(id: u32) -> ContractId {
+    ContractId::from_low_u64_be(id as u64)
+}
+
+pub fn command_topic(id: ContractId) -> Vec<u8> {
+    format!("phala/contract/{}/command", hex::encode(&id)).as_bytes().to_vec()
```

Comment:
Do we assume commands are encrypted, while events are not?

Reply:
> Do we assume commands are encrypted, while events are not?

Exactly